core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -33,12 +33,87 @@ import fs2.internal._
+  * _Note_ that type `Nothing` is polykinded type, so it can also be
+  * applied as an argument to the type parameter `F[_]`.
+  * Nevertheless, for the sake of clarity, in `fs2` we use the type alias
+  * `type Pure[A] = Nothing` to represent a pure pull as `Pull[Pure, O, R]`.
+  *
+  * A '''silent''' pull is one known to emit no outputs. In a silent pull, the
+  * output type parameter `O` is bound to `Nothing`, the bottom type.
+  *
+  * A '''broken''' pull is known never to terminate successfully, which may happen
+  * because the pull is known to fail, or is known to be interrupted.
+  * {{{
+  * def infinite: Pull[Pure, Nothing, Nothing] = F.unit >> infinite
+  * }}}
+  * A broken pull may also be one that contains an action in the effect F
+  * that fails or is known not to terminate, such as `IO.never` from `cats.effect`.
+  *
+  * A '''closed''' pull is one that, if may terminate successfully with no
+  * carry-over information. The type of a closed pull has the type parameter `R`
+  * bound to `Unit`. A `Stream[F, O]` is just a wrapper  around a closed pull.
+  *
+  * === Evaluation and Compilation ===
+  *
+  * Like functional effect types (e.g. `IO` in Cats Effect),   a pull is just
+  * a passive data structure that describes a sequence of action.
+  * It is neither a running process nor a container to be eventually filled
+  * with the  result of a concurrent process (like `scala.concurrent.Future`).
+  *
+  * To "run" or "interpret" a pull, it has to be compiled.
+  * A '''closed''' pull `p: Pull[F, O, Unit]`  is compiled into a single
+  * effectful action `q: F[S]`.
+  *
+  * The effectful action `q` is a combination, via the monad instance of `F`,
+  * of all the actions in the effect `F` that come out of the pull.
+  * The result of that `F` action is a value of type `S`, that is the result
+  * of combining the outputs emitted by the pull, in the order it emits them,
+  * using a _foldMap_ function. Depending on that function, it may collect
+  * all outputs into a list (or Vector or Array), add them up together into
+  * a single value, or just discard them altogether (by _draining_ the pull).
+  *
+  * Compilation is pull-based, rather than push-based (hence the name of the datatype).
+  * It is the compilation process itself, that determines when the compilation
+  * of each single effect can proceed or is hold back. Effects and outputs later
+  * in the pull are not performed or emitted, _unless and until_ the compilation
+  * calls for them.
+  *
+  * === Resource scoping ===
+  *
+  * The side effects performed by a pull may operate on resources. These resources
+  * need to be 1. retained during the execution of the pull (compilation),
+  * 2. shared by several pulls that we may want to unroll and evaluate independently,
+  * and 3. cleaned up once they are no longer needed.
+  * To keep track of resources, pull compilation uses resource scopes: collection
```

Comment:
I don't think we should mention compilation here. Scopes are part of the overall API and don't need to be described in terms of how they are used during compilation.